Refactored NuGet.Configuration

PatoBeltran opened the pull request:
Improved API for parsing nuget.config by creating an object model that is easily extensible and updated current API usage to conform with the new usage patterns.

Previously, the nuget.config logic was centered around Settings.cs, which was in charge of loading all nuget.configs, figuring out how to merge them and giving an API to the user to interact with the content of the config files. Because of this, the usage was limited for interacting with nuget.config. The API was too specific for parsing only a few elements and enabling new elements was difficult because this meant adding a bunch of APIs specific for any new element.

This PR adds an abstraction of a single nuget.config file (SettingsFile.cs), with its own abstraction of its content (NugetConfiguration.cs), making it easy to parse and add new elements. Then the Settings object is only in charge of finding and loading every single config file and merging their settings together. The settings file give the user a way to interact with the computed values, instead of with the values from the config files directly.

The users of the APIs were also updated to follow the new pattern. PackageSourceProvider was the one with the heaviest usage of the API and was rewritten to follow the principles of the new API.

Given that this is a big PR, here are the most important files, be sure to pay special attention to all of the files in this list, most of the changes in other files are just updates to use the new API instead of the old one:

**Setting file parsing**

SettingBase.cs 
SettingText.cs 
SettingElement.cs 
ISettingsGroup.cs 
SettingsGroup.cs 
NuGetConfiguration.cs 
SettingSection.cs 
ParsedSettingSection.cs 
MergedSettingSection.cs 
SettingItem.cs 
AddItem.cs 
ClearItem.cs  
CredentialsItem.cs 
SourceItem.cs
UnknownItem.cs 

SettingFactory.cs 

**Settings API**

ISettings.cs
Settings.cs
SettingsFile.cs
SettingsUtility.cs

**Important changes in the way it uses the settings API**

IPackageSourceProvider.cs 
PackageSourceProvider.cs 


**Tests: Settings File Parsing**

SettingsFileParsingTests/SettingsFileTests.cs 
SettingsFileParsingTests/NuGetConfigurationTests.cs 
SettingsFileParsingTests/SettingSectionTests.cs 
SettingsFileParsingTests/AddItemTests.cs 
SettingsFileParsingTests/SourceItemTests.cs 
SettingsFileParsingTests/ClearItemTests.cs 
SettingsFileParsingTests/CredentialsItemTests.cs 

**Tests: Settings API**

SettingsTests.cs 
SettingsUtilityTests.cs 

**Tests: Important changes in the way it uses the settings API**

PackageSourceProviderTests.cs

PatoBeltran committed:
Refactored NuGet.Configuration

Improved API for parsing nuget.config by creating an object model that
is easily extensible and updated current API usage to conform with the
new usage patterns.

PatoBeltran committed:
fixed functional tests

PatoBeltran committed:
add missing tests

PatoBeltran committed:
fix linux tests

PatoBeltran commented on src/NuGet.Core/NuGet.Configuration/Settings/Nodes/SettingsSection.cs:
```diff
@@ -0,0 +1,158 @@
+namespace NuGet.Configuration
+{
+    // This could also be an abstract generic class where T is a settingsItem if we wanted to constraint each section to an specific type of item
+    public sealed class SettingsSection : SettingsCollection<SettingsItem>, IEquatable<SettingsSection>
```
Currently, this class is a generic Section class for all sections... This class could be made abstract and extend a class for each of the currently known and used sections, that way we could specify rules for each of the sections (for example, let `packageSources` be cleared but not `config`). @NuGet/nuget-client do you think this is worth it? or should we just keep it as a general Section element?

PatoBeltran committed:
Fixed spacing and config command message

zhili1208 commented:
looks like ISettings API is totally changed, I believe lots of tools which are using nuget library depends on that API. two questions here 1. do we have to totally change the interface? 2. If we have to do that, do we have any plan which can make user migrate their code to new nuget version easily?

PatoBeltran committed:
PackageSourceProvider.SavePackageSources should be able to update credentials

PatoBeltran committed:
Add deprecated APIs back and mark them as deprecated

PatoBeltran committed:
fix build

PatoBeltran commented:
@zhili1208 I added back the old API in `ISettings` and marked them as obsolete. The only three things that I removed from `ISettings` where `Priority`, `Root` and `Filename`. `Root` and `Filename` do not make sense to be part of `ISettings` since `ISettings` refers to a collection of computed config files, not just one. And `Priority` seems to be something that should not be public in the `Settings` implementation. In `SettingsUtility` there are methods to get the roots and file paths of all the configs used in a given `ISettings` object. Do you think this is an acceptable modification of the `ISettings` API?

PatoBeltran committed:
element attributes can be empty strings

PatoBeltran committed:
Add support for ValidAuthenticationTypes

PatoBeltran committed:
set value with null should delete existing values

rohit21agrawal commented:
why can't we leave ISettings as is for backward compat, and make a new one like ISettings2 and use it everywhere in our code?

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/ISettings.cs:
```diff
@@ -11,52 +11,41 @@ namespace NuGet.Configuration
-        /// <summary>
-        /// Folder under which the config file is present
-        /// </summary>
-        string Root { get; }
```
similar to APIs there might be usage of these members as well, so why not mark them obsolete as well?

PatoBeltran replied:
`ISettings` used to be not only the computed settings but also a single nuget.config, therefore it has a pointer to the `root` and `filename`. In the new design, `ISettingsFile` is the pointer to a single nuget.config, but `ISettings` only manages the computed values. If we only mark them as deprecated, what values should the implementation return?

jainaashish replied:
`ISettings used to be not only the computed settings but also a single nuget.config, therefore it has a pointer to the root and filename.` Have the same value even with new design...

PatoBeltran replied:
`ISettings` used to point to **one specific** nuget.config, and had pointers to the next nuget.config, thus depending on which `ISettings` you called you could have gotten a different setting value (most of out higher level APIs exposed only one of this configs). The new `ISettings` points to **n** nuget.config (i.e. the computed settings), therefore there is no reference to "the same value" as the previous design. What `root` and `filename` of which config do you think we should be showing if we left this option?

If we are really concerned about deleting this, we could mark them obsolete with `error = true` and have the implementation just return an empty string, but that would have the same "breaking" effect for anyone that is actually using this (which I don't expect to be anyone, since `SettingsUtility` has utility methods to extract the `root` or `filename` of all `SettingsFile` for a given `ISetting`)

jainaashish replied:
Discussed offline, if @rrelyea is fine with breaking existing usage of these interface in 15.9 then I'm fine too.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/ISettings.cs:
```diff
@@ -113,15 +108,47 @@ public interface ISettings
+        /// <summary>
+        /// Adds or updates the given <paramref name="item"/> to the settings.
+        /// If the <paramref name="item"/> has to be added this method will add it
+        /// in the furthest compatible settings file.
```
can you elaborate furthest compatible settings file?

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/ISettingsFile.cs:
```diff
@@ -0,0 +1,52 @@
+        /// </summary>
+        NuGetConfiguration RootElement { get; }
+
+        bool IsDirty { get; set; }
```
missing documentation

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/ISettingsFile.cs:
```diff
@@ -0,0 +1,52 @@
+    /// <summary>
+    /// Interface to expose a NuGet settings file
+    /// </summary>
+    internal interface ISettingsFile
```
I'd imagine the define is each ISettingFile manage it's own nuget.config then why doesn't this interface expose functionality to add/update data to this setting file?

PatoBeltran replied:
You can interact with the elements directly because the `RootElement` is exposed. I could add a `SetItemInSection` to have a similar behavior than `ISettings`, but I thought it was enough to expose the elements and let the user interact with them directly. This is an `internal` interface, therefore the only users should be inside `NuGet.Configuration`.

jainaashish replied:
Why would you like to expose `RootElement` instead of let each SettingFile manages their own content? And that'd also help if tomorrow we want to expose individual setting files in UI or dynamic reloading of config files.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/ISettingsFile.cs:
```diff
@@ -0,0 +1,52 @@
+        /// <summary>
+        /// Root element for the settings abstraction
+        /// </summary>
+        NuGetConfiguration RootElement { get; }
```
why should it expose `RootElement`? shouldn't SettingFIle handle all manupulation for `NuGetConfiguration` since this a wrapper over it?

PatoBeltran replied:
I think handling all the manipulation from this class is limiting. At the end of the day, the user of this interface knows how the data is stored and how to handle it. If we add specific methods for interacting with the config files here we would be constrained to these APIs (the same way we were constraint by the previous APIs in `ISettings`). This is the reason I opted for exposing the root element and letting the user handle the data directly. Also, note that this is an internal API, therefore the user should know how the data works. 

Any non-internal user should always interact with the computed settings through `ISettings` instead of `ISettingsFile`.

jainaashish replied:
`If we add specific methods for interacting with the config files here we would be constrained to these APIs (the same way we were constraint by the previous APIs in ISettings).`

Not really, what i'm asking is to have apis similar to `ISettings` to manage sections which isn't constraint by current data.

`This is the reason I opted for exposing the root element and letting the user handle the data directly`

This is the part I dont understand, if we have already defined an interface to interact with individual setting file then why would I expose an element to directly handle the underneath data?

`Also, note that this is an internal API, therefore the user should know how the data works.`

This has nothing to do with either internal or public as long as there are multiple layers and each layer knows what it's suppose to do. In current code, It feels like there are multiple layers but functionality and responsibilities are not distributed among them... Also, if you properly handle all the data manipulation at `ISettingFile` then tomorrow exposing individual config files in UI, will be far more convenient and doable.

PatoBeltran replied:
My initial thought was that adding an API to manage this would not be necessary since the providers that could call directly into this will know how to handle the API, but I can add the following methods to `ISettingsFile` and not expose the root element so that the API is consistent with `ISettings`:

```
SettingsSection GetSection(string sectionName); 

bool CreateSection(SettingsSection section, bool isBatchOperation = false); 

bool SetItemInSection(string sectionName, SettingsItem item, bool isBatchOperation = false); 
```

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs:
```diff
@@ -46,96 +42,174 @@ public class Settings : ISettings
+            if (_computedSections.TryGetValue(sectionName, out var section))
             {
-                throw new ArgumentException(Resources.Settings_FileName_Cannot_Be_A_Path, nameof(fileName));
+                if (section.TryUpdateChildItem(item, isBatchOperation))
```
How this change propagated to corresponding `SettingFile`? Does `SettingSection` and `SettingFile` also interact with each other? I hope not... SettingFile and SettingSection should be independent of each other.

PatoBeltran replied:
Each item has a reference to its parent element (which is in the tree of a specific `SettingsFile`), so each item will update itself and the corresponding `SettingsFile` and then the computed settings will be recomputed.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs:
```diff
@@ -46,96 +42,174 @@ public class Settings : ISettings
+            if (_computedSections.TryGetValue(sectionName, out var section))
             {
-                throw new ArgumentException(Resources.Settings_FileName_Cannot_Be_A_Path, nameof(fileName));
+                if (section.TryUpdateChildItem(item, isBatchOperation))
```
Shouldn't you just return here irrespective update is a success or failure?

PatoBeltran replied:
It can be the case that the given section doesn't have that item so this will return `false`. In any case, if we were unable to update the child we then want to try to add it, not just fail the operation.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs:
```diff
@@ -46,96 +42,174 @@ public class Settings : ISettings
+
+            if (outputSettingsFile != null)
+            {
+                return outputSettingsFile.RootElement.AddChild(section, isBatchOperation);
```
how are you updating your `_computedSections` to add this?

PatoBeltran replied:
On any edit to a specific `ISettingsFile` and event is triggered and `_computedSections` is recalculated.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs:
```diff
@@ -46,96 +42,174 @@ public class Settings : ISettings
+            var outputSettingsFile = GetOutputSettingsFileForSection(sectionName);
+
+            // Set is an add
+            return outputSettingsFile != null && outputSettingsFile.RootElement.AddItemInSection(sectionName, item.Copy(), isBatchOperation);
```
how does this get refreshed in `_computedSections`?

PatoBeltran replied:
On any edit to a specific `ISettingsFile` and event is triggered and `_computedSections` is recalculated.

jainaashish replied:
ahh I see, dont you think it might be little over-do? Since each time, you need to handle a event to merge all the setting files, where you already know which file has changed.

Also, would it be possible to update multiple setting files at the same time? If not, then you can go for simple solution.

PatoBeltran replied:
We do know what file has changed, but we don't know how those changes could affect the final computed settings until we recompute everything. Think of what if the change means that a clear was removed, or that a setting is added to the first config but there is already a setting closer to the user that will override this (thus nothing would change in the computed settings). I think that to store all the state to know exactly what changed and how to update the computed settings would make this code really complicated. If you have a simpler solution which is more efficient I would love to hear it.

To answer your question about editing multiple configs, every update operation has a flag (`isBatchOperation`), which tells the method if Save() should be called in the config that changed. Basically, if you are editing multiple settings (updates, adds, removes) you can specify this in every operation and save until the end. If you are only messing with `ISettings` (i.e computed settings), the update operations will update the data in memory, and saving will flush the data to the files.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs:
```diff
@@ -46,96 +42,174 @@ public class Settings : ISettings
+            var curr = _settingsHead;
+            while (curr != null)
+            {
+                curr.SettingsChanged += (_, __) => { RefreshSettingValues(); };
```
do you really need to rely on `SettingsChanged` event from `SettingFile` to take latest changes in `ISettings`? If I'm not wrong then every change in any of the `SettingFile` happens through `ISettings`...

PatoBeltran replied:
As I just focused this refactor in getting the same behavior as we had previously, then all the changes happen from `ISettings`, but this is not true for the whole API. The API has the capability for users only editing one `ISettingsFile` directly (e.g. if we add the ability to edit specific configs in the UI), therefore I made it so that if there is any change with any of the `ISettingsFile` used in a given `ISettings` then we recompute the merged values.

The previous code calculated the merged values in every `Get` operation, I updated this to make `get` more performant (since we do way more gets than sets), and only recompute the values on each `set`. Therefore I would say that this code is way more performant than the previous.

jainaashish replied:
Even if we allow editing individual config file through UI, that will also be through `Settings` or would we allow editing multiple files at the same time? 

I think we can take all this discussion offline and finalize it...?

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs:
```diff
@@ -46,96 +42,174 @@ public class Settings : ISettings
+            var computedSections = new Dictionary<string, SettingsSection>();
+
+            var curr = _settingsHead;
+            while (curr != null)
```
As I mentioned above... this adds unnecessary additional perf cost of merge all the settings files over and over again each time you change a single item in any of the file.

PatoBeltran replied:
I don't believe this perf cost is unnecessary. This is only done on each set operation (i.e. when a settings file is saved). By doing this we gain the freedom in the API to interact with either computed settings or directly with a settings file without having a wrapper in between where you are limited to the exposed interface of that given wrapper. 

The previous code computed and merged all settings files *on every get operation*, this update changes that to compute everything on sets because set operations are way less common than get operations.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs:
```diff
@@ -46,96 +42,174 @@ public class Settings : ISettings
+            var clearedSections = writteableSettingsFiles.Select(s => s.RootElement.GetSection(sectionName)).Where(s => s != null && s.HasClearChild());
+            if (clearedSections.Any())
+            {
+                return clearedSections.Last().Origin;
```
It seems your `SettingSection` is doing all the magic... :) But it's really hard to read... not sure if this was the existing code but anything we can do to improve readability?

PatoBeltran replied:
What do you mean? This method is finding the "most compatible" config file to add something in a given section. I agree that this looks like a hack, but this is only to keep the current behavior of how to write things in configs, I'm happy to discuss this again if you don't agree about what config we should be adding things for each section.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs:
```diff
@@ -46,96 +42,174 @@ public class Settings : ISettings
+        private void RefreshSettingValues()
+        {
+            MergeSettingFilesValues();
+            SettingsChanged?.Invoke(this, EventArgs.Empty);
```
`SettingsChanged` event should not be raised here since you haven't saved the file yet.

PatoBeltran replied:
This method is only called from the `SettingsChanged` of a specific `SettingsFile`, I could remove this method and only make it inline in the event handler so that no one can call this without changing a file.

jainaashish reviewed (commented):
Not yet done, will continue to review and post my feedback.

PatoBeltran committed:
Fix tests and address feedback

jainaashish commented:
I discussed it offline with Pato, and we're clearly on the same page in terms of design and implementation. I feel like there should have been a clear design discussion before making lot of assumptions. But never too late, I propose we discuss it as a team and decide the right way forward so that we can quickly and effectively make progress here.

@PatoBeltran can you please set up some time to finalize the design and behavior?

cc @NuGet/nuget-client

PatoBeltran committed:
Updated API with single point of entry

PatoBeltran commented:
@NuGet/nuget-client I updated the PR with the feedback that I received. The whole API has only one point of interaction (i.e. `ISettings`), whenever a user gets a hold of a section or an item this is a copy of the actual abstraction, so modifying it won't cause settings to change unless it is passed through `ISettings` methods, which won't save to disk until the user explicitly calls `ISettings.SaveToDisk()`

The only open question left is: should the `settingsChanged` event be fired once the In-memory abstraction of settings changed or when the disk changed? Currently, it works with the later, but it can be easily changed.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/ISettings.cs:
```diff
@@ -113,15 +108,51 @@ public interface ISettings
+        /// <param name="sectionName">section where the <paramref name="item"/> has to be added. If this section does not exist, one will be created.</param>
+        /// <param name="item">item to be added to the settings.</param>
+        /// <returns>true if the item was successfully updated or added in the settings</returns>
+        void AddOrUpdate(string sectionName, SettingItem item);
```
change the method return type to bool

PatoBeltran replied:
I explicitly made this void because I wanted to give more info to the user in the case it didn't succeed. If the add or update failed is because it is trying to update a machine-wide config, therefore I want the user to have this information instead of just returning false and ignoring the changes.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/ISettings.cs:
```diff
@@ -113,15 +108,51 @@ public interface ISettings
+        /// <param name="sectionName">Section where the <paramref name="item"/> is stored. If this section does not exist, the method will throw</param>
+        /// <param name="item">item to be removed from the settings</param>
+        /// <remarks> If the <paramref name="item"/> is in a machine wide config this method will throw</remarks>
+        void Remove(string sectionName, SettingItem item);
```
dow e also neede `removeSection` api?

PatoBeltran replied:
We could add it, the same effect would happen if you remove all the items in a section. I don't think there is any scenario where removing the whole section makes sense, that's why I didn't think to add it.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/ISettings.cs:
```diff
@@ -113,15 +108,51 @@ public interface ISettings
+        /// <summary>
+        /// Flushes any update that has been done in memory through the ISettings API to the settings file in disk.
+        /// </summary>
+        void SaveToDisk();
```
so does it mean that there is no batch or non-batch modes now?

PatoBeltran replied:
Every update messes with the in-memory abstraction only and the user has to explicitly save to get it in disk. This is the same pattern followed by the dotnet configuration API.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs:
```diff
@@ -46,96 +43,209 @@ public class Settings : ISettings
+            if (_computedSections.TryGetValue(sectionName, out var section))
             {
-                throw new ArgumentException("root cannot be null or empty", nameof(root));
+                section.Add(item);
```
what if item already exists in computedSection, odes it also update? if yes, then may be change it to AddorUpdate?

PatoBeltran replied:
Since `_computedSection` only has pointers to the items that are held in the `settingsFile` the `AddOrUpdate` above this line takes care of the update scenario. I will add some comments to clarify this.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs:
```diff
@@ -46,96 +43,209 @@ public class Settings : ISettings
+
+            // Operation is an add
+            var outputSettingsFile = GetOutputSettingFileForSection(sectionName);
+            if (outputSettingsFile != null)
```
when would it not get outputSettingFile? and if there are cases like those, then may be throw appropriate error..?

PatoBeltran replied:
We could throw. This basically means that if there is no writeable file we ignore the update. I opted to ignore instead of throw to keep the same behavior as before, but I'm in favor of throwing to have a better error experience.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/Settings.cs:
```diff
@@ -46,96 +43,209 @@ public class Settings : ISettings
             }
 
-            if (!FileSystemUtility.IsPathAFile(fileName))
+            Debug.Assert(item.IsAbstract());
```
remove these Debug Asserts

PatoBeltran replied:
Why? I think they give good documentation of what is expected to have on those methods at that given point.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingsFile.cs:
```diff
@@ -0,0 +1,240 @@
+                });
+
+                IsDirty = false;
+                SettingsChanged?.Invoke(this, EventArgs.Empty);
```
shouldn't this SettingChanged event be raised only once per `SaveToDisk` in `Settings.cs`?

PatoBeltran replied:
You could subscribe to this `SettingsChange` if you want to know if a specific settingsFile changed. Settings subscribe to this event on each settings file. You are right that on settings I need to make sure this event shouldn't trigger the Settings event more than once. Good catch!

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingsFile.cs:
```diff
@@ -0,0 +1,240 @@
+        internal bool TryGetSection(string sectionName, out SettingSection section)
+        {
+            var result = _rootElement.Sections.TryGetValue(sectionName, out var parsedSection);
+            section = parsedSection;
```
you can directly use `parsedSection` while reading from rootElement.

PatoBeltran replied:
the compiler complained about this because of the types, that's why I had to have an extra variable.

jainaashish commented on src/NuGet.Core/NuGet.Configuration/Settings/AbstractSettingSection.cs:
```diff
@@ -0,0 +1,158 @@
+
+        internal override SettingBase Clone()
+        {
+            return new AbstractSettingSection(Name, Attributes, Items.Select(s => s.Clone() as SettingItem));
```
Wasn't able to follow `Attributes` origin but just make sure that you're doing proper cloning.

PatoBeltran replied:
In our current schema, there are no sections with attributes... This refactor enables the possibility to add attributes in sections... sections inherit from SettingElement which is the one that manages the attributes, there's where this variable comes from.

jainaashish reviewed (approved):
Great job Pato 👍 this looks great. Although I've not gone through all the file changes but the important ones and looks good to me.

PatoBeltran committed:
fix tests

PatoBeltran committed:
address comments

PatoBeltran committed:
fix unit test

nkolev92 commented on src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx:
```diff
@@ -458,8 +458,7 @@ nuget update -Self</value>
   <data name="Option_ConfigFile" xml:space="preserve">
-    <value>The NuGet configuration file. If not specified, file %AppData%\NuGet\NuGet.config is used as configuration file.</value>
-    <comment>Please don't localize "%AppData%\NuGet\NuGet.config".</comment>
+    <value>The NuGet configuration file. If specified, only the settings from this file will be used. If not specified, the hierarchy of configuration files  from the current directory will be used. To learn more about nuget configuration go to https://docs.microsoft.com/en-us/nuget/consume-packages/configuring-nuget-behavior</value>
```
I think we should setup a fw link here. More future proof. 
PMs should be able to help with that.

PatoBeltran replied:
cc. @anangaur

nkolev92 commented on src/NuGet.Clients/NuGet.CommandLine/Commands/ConfigCommand.cs:
```diff
@@ -1,4 +1,4 @@
-﻿using System;
```
revert this file.

nkolev92 replied:
Actually this one is missing a header, so maybe add a header :)

nkolev92 commented on src/NuGet.Clients/NuGet.CommandLine/Commands/RestoreCommand.cs:
```diff
@@ -365,7 +365,7 @@ private async Task<RestoreSummary> PerformNuGetV2RestoreAsync(PackageRestoreInpu
                 return new RestoreSummary(
                     result.Restored,
                     "packages.config projects",
-                    Settings.Priority.Select(x => Path.Combine(x.Root, x.FileName)),
```
Did you verify that the configs are printed in the same order when you do restores before/after this change.

PatoBeltran replied:
Yes, this does not changes

dtivel replied:
Maybe add a test verifying order?

nkolev92 commented on src/NuGet.Clients/NuGet.CommandLine/Commands/SourcesCommand.cs:
```diff
@@ -83,62 +87,66 @@ public override void ExecuteCommand()
-            var sourceList = SourceProvider.LoadPackageSources().ToList();
-            var existingSource = sourceList.Where(ps => String.Equals(Name, ps.Name, StringComparison.OrdinalIgnoreCase));
-            if (!existingSource.Any())
+            var packageSource = SourceProvider.GetPackageSourceWithName(Name);
```
Nit: ByName?

nkolev92 commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/VSSettings.cs:
```diff
@@ -91,96 +91,118 @@ private void OnSolutionOpenedOrClosed(object sender, EventArgs e)
         }
 
-        public bool DeleteValue(string section, string key)
+        public void AddOrUpdate(string sectionName, SettingItem item)
```
Why not keep the bools for op result?

PatoBeltran replied:
I opted to use void and throw in the case where there is a failure, this way we can give more information to the user as of why the operation failed. Most of the times this failure is because there is no writeable config or because there was some bug in the implementation, throwing will give more context about what is happening.

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/NuGet.Configuration.csproj:
```diff
@@ -47,6 +47,6 @@
     </EmbeddedResource>
   </ItemGroup>
 
-  <Import Project="$(BuildCommonDirectory)common.targets"/>
```
revert this.

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/PackageSource/IPackageSourceProvider.cs:
```diff
@@ -8,17 +8,109 @@ namespace NuGet.Configuration
 {
     public interface IPackageSourceProvider
     {
+        //TODO: Delete all obsolete APIs.
```
Every TODO should be accompanied by an issue link.

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceCredential.cs:
```diff
@@ -185,5 +186,46 @@ private static IEnumerable<string> ParseAuthTypeFilterString(string str)
+            }
+
+            return string.Equals(Source, other.Source, StringComparison.OrdinalIgnoreCase) &&
+                string.Equals(Username, other.Username, StringComparison.OrdinalIgnoreCase) &&
```
username and password are not OrdinalIgnoreCase.

PatoBeltran replied:
@nkolev92 is the source OrdinalIgnoreCase?

nkolev92 replied:
http feeds are, but file feeds are not. 
So I'd guess it's no? But I'd do some digging

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs:
```diff
@@ -298,178 +240,472 @@ private PackageSourceCredential ReadCredentialFromEnvironment(string sourceName)
+                        isDirty = true;
                     }
-                    hasChanges = true;
+                    catch { }
```
I don't understand this. In which cases do we throw/why do we swallow the exception.

PatoBeltran replied:
This is to preserve previous behavior... This will fail when the user is trying to remove a machine-wide source. The previous behavior was to swallow any error and just ignore the user's request. Since this PR is focused on config changes and not sources changes, I didn't make any functional changes to this class.

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs:
```diff
@@ -298,178 +240,472 @@ private PackageSourceCredential ReadCredentialFromEnvironment(string sourceName)
+                    addedInSameFileAsCurrentSource = true;
+                }
+                // We ignore any errors since this means the current source file could not be edited
+                catch { }
```
Similar to the above, why swallow the exception here?

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/Properties/AssemblyInfo.cs:
```diff
@@ -1,6 +1,15 @@
 
 [assembly: CLSCompliant(true)]
+
+#if SIGNED_BUILD
```
👏 
Wasn't aware that we do this. 
Have we done it in any other projects?

PatoBeltran replied:
There are a couple, but we should do it more to have a more clearly defined public API.

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/Resources.resx:
```diff
@@ -120,9 +120,26 @@
     <value>Value cannot be null or empty string.</value>
   </data>
+  <data name="CannotUpdateDifferentItems" xml:space="preserve">
+    <value>The item passed to the Update method cannot refer to a different item to the one to be updated.</value>
```
Consider
> The item passed to the Update method cannot refer to a different item ~~to~~ **than** the one ~~to be~~ **being** updated.

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/Resources.resx:
```diff
@@ -120,9 +120,26 @@
+    <value>Unable to update setting since it is in a machine wide NuGet.Config</value>
+  </data>
+  <data name="Error_CouldNotDisableSources" xml:space="preserve">
+    <value>Could not add disabled package source</value>
```
nit: add . to all the messages.

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/Resources.resx:
```diff
@@ -132,18 +149,35 @@
     <value>Environment variable '{0}' must contain an absolute path, the full path of '{1}' cannot be determined.</value>
   </data>
+  <data name="SectionDoesNotExist" xml:space="preserve">
+    <value>The section does not exist in the settings</value>
```
should we saty which section?

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/Settings/AbstractSettingSection.cs:
```diff
@@ -0,0 +1,158 @@
+        }
+
+        // This is public for testing
+        public AbstractSettingSection(string name, params SettingItem[] children)
```
you have internalsVisibleTo, doesn't need to be public.

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/Settings/AbstractSettingSection.cs:
```diff
@@ -0,0 +1,158 @@
+
+namespace NuGet.Configuration
+{
+    public sealed class AbstractSettingSection : SettingSection
```
I'm confused as to what this AbstractSettingSection represents.

PatoBeltran replied:
There are two types of sections... the ones that are actual sections that were read from a file and have a pointer to a settings file and the ones that are a collection of computer values... I decided to separate this two in `AbstractSettingSection` and `ParsedSettingSection` to have the code be a little easier to read, since the abstract sections are only an in-memory representation (i.e. don't map to one specific section in a file), but the parsed sections have a counterpart in a file.

Do you think this is not clear enough? what would you propose to make it better?

PatoBeltran replied:
@nkolev92 do you think `InMemorySettingSection` is a better name?

dtivel replied:
`VirtualSettingSection`?  `MergedSettingSection`?

nkolev92 commented on src/NuGet.Core/NuGet.Configuration/Utility/SettingsUtility.cs:
```diff
@@ -12,21 +12,55 @@ namespace NuGet.Configuration
     public static class SettingsUtility
     {
-        public static readonly string ConfigSection = "config";
+        //TODO: Delete all obsolete APIs.
```
Link the issue and TODO.

nkolev92 reviewed (approved):
Mostly minor stuff and only one that absolutely has to change (the credentials casing thing).

Did as much as much I can :)
Lots of great improvements here, and the API changes are as minimal as possible.

PatoBeltran committed:
add missing tests

PatoBeltran committed:
Address feedback

PatoBeltran committed:
change setting element Name to ElementName

PatoBeltran committed:
Do not explicitly constraint a list of allowed attributes for existing items

dtivel commented on src/NuGet.Clients/NuGet.CommandLine/Commands/SourcesCommand.cs:
```diff
@@ -223,26 +225,25 @@ private void UpdatePackageSource()
-            bool authTypesEmpty = String.IsNullOrEmpty(ValidAuthenticationTypes);
+            var usernameEmpty = string.IsNullOrEmpty(Username);
+            var passwordEmpty = string.IsNullOrEmpty(Password);
+            var authTypesEmpty = string.IsNullOrEmpty(ValidAuthenticationTypes);
```
`isUsernameEmpty`, `isPasswordEmpty`, and `isAuthTypesEmpty`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs:
```diff
@@ -298,178 +240,472 @@ private PackageSourceCredential ReadCredentialFromEnvironment(string sourceName)
+        {
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentNullException(nameof(name));
```
This is incorrect.  If you were only checking if `name` is `null` then it would be appropriate to throw an `ArgumentNullException`; however, because you're checking for both `null` and a zero-length string, you either need to throw two separate exceptions or throw one exception that is appropriate for both cases.  I suggest the latter option as it's simpler:
```
if (string.IsNullOrEmpty(name))
{
    throw new ArgumentException(Resources.Argument_Cannot_Be_Null_Or_Empty, nameof(name));
}
```
Like [here](https://github.com/NuGet/NuGet.Client/pull/2370/files#diff-f83497230965c941a0bf5e2d2ed13deeR134).

This feedback applies throughout the PR.

dtivel commented on src/NuGet.Clients/NuGet.CommandLine/NuGetCommand.resx:
```diff
@@ -458,8 +458,7 @@ nuget update -Self</value>
   <data name="Option_ConfigFile" xml:space="preserve">
-    <value>The NuGet configuration file. If not specified, file %AppData%\NuGet\NuGet.config is used as configuration file.</value>
-    <comment>Please don't localize "%AppData%\NuGet\NuGet.config".</comment>
+    <value>The NuGet configuration file. If specified, only the settings from this file will be used. If not specified, the hierarchy of configuration files  from the current directory will be used. To learn more about nuget configuration go to https://docs.microsoft.com/en-us/nuget/consume-packages/configuring-nuget-behavior</value>
```
* Remove extra space in "...files  from..."
* Properly capitalize NuGet in "To learn more about nuget configuration..."
* Terminate the last sentence with a period.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.PowerShellCmdlets/PackageActionBaseCommand.cs:
```diff
@@ -409,9 +410,9 @@ protected virtual DependencyBehavior GetDependencyBehavior()
+            var dependencySetting = SettingsUtility.GetConfigValue(ConfigSettings, "dependencyversion");
             DependencyBehavior behavior;
-            bool success = Enum.TryParse(dependencySetting, true, out behavior);
+            var success = Enum.TryParse(dependencySetting, true, out behavior);
```
Add an `ignoreCase:` named parameter to the mysterious literal `true`.

dtivel commented on src/NuGet.Clients/NuGet.CommandLine/CommandLineMachineWideSettings.cs:
```diff
@@ -1,4 +1,4 @@
-﻿using System;
+using System;
```
License header?

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.PowerShellCmdlets/PackageActionBaseCommand.cs:
```diff
@@ -409,9 +410,9 @@ protected virtual DependencyBehavior GetDependencyBehavior()
         protected DependencyBehavior GetDependencyBehaviorFromConfig()
         {
-            string dependencySetting = ConfigSettings.GetValue("config", "dependencyversion");
+            var dependencySetting = SettingsUtility.GetConfigValue(ConfigSettings, "dependencyversion");
```
Should `dependencyversion` be a constant?  I see multiple instances of the string literal, which usually implies yes.

`ConfigurationConstants` seems like a good place for it.

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -293,9 +294,11 @@ private static bool IsUILegalDisclaimerSuppressed()
+
                 DependencyBehavior behavior;
-                var success = Enum.TryParse(dependencySetting, true, out behavior);
+                var success = Enum.TryParse(dependencySetting?.Value, true, out behavior);
```
`ignoreCase:`

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/PackageManagerControl.xaml.cs:
```diff
@@ -293,9 +294,11 @@ private static bool IsUILegalDisclaimerSuppressed()
             {
-                var dependencySetting = nugetSettings.GetValue("config", "dependencyversion");
+                var configSection = nugetSettings.GetSection(ConfigurationConstants.Config);
+                var dependencySetting = configSection?.GetFirstItemWithAttribute<AddItem>(ConfigurationConstants.KeyAttribute, "dependencyversion");
```
Const-ize `dependencyversion`?

dtivel commented on src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/VSSettings.cs:
```diff
@@ -91,96 +91,118 @@ private void OnSolutionOpenedOrClosed(object sender, EventArgs e)
+#pragma warning disable CS0618 // Type or member is obsolete
 
-        public string Root => SolutionSettings.Root;
+        [Obsolete("GetValue(...) is deprecated, please use GetSection(...) to interact with the setting values instead.")]
```
In all cases:

    "...is deprecated. Please..."

dtivel commented on src/NuGet.Clients/NuGet.VisualStudio.Common/BindingRedirectBehavior.cs:
```diff
@@ -29,29 +25,43 @@ public bool IsSkipped
             get
             {
-                string settingsValue = _settings.GetValue(BindingRedirectsSection, SkipBindingRedirectsKey) ?? string.Empty;
+                var bindingRedirectSection = _settings.GetSection(ConfigurationConstants.BindingRedirectsSection);
```
`bindingRedirectsSection`

dtivel commented on src/NuGet.Clients/NuGet.VisualStudio.Common/BindingRedirectBehavior.cs:
```diff
@@ -29,29 +25,43 @@ public bool IsSkipped
-                _settings.SetValue(BindingRedirectsSection, SkipBindingRedirectsKey, value.ToString(CultureInfo.InvariantCulture));
+                _settings.AddOrUpdate(ConfigurationConstants.BindingRedirectsSection,
+                    new AddItem(ConfigurationConstants.SkipBindingRedirectsKey, value.ToString(CultureInfo.InvariantCulture)));
+                _settings.SaveToDisk();
```
Did the old code implicitly save to disk on set?

PatoBeltran replied:
Yes

dtivel commented on src/NuGet.Clients/NuGet.VisualStudio.Common/BindingRedirectBehavior.cs:
```diff
@@ -29,29 +25,43 @@ public bool IsSkipped
             get
             {
-                string settingsValue = _settings.GetValue(BindingRedirectsSection, FailOnBindingRedirects) ?? string.Empty;
+                var bindingRedirectSection = _settings.GetSection(ConfigurationConstants.BindingRedirectsSection);
```
`bindingRedirectsSection`

dtivel commented on src/NuGet.Clients/NuGet.VisualStudio.Common/PackageManagementFormat.cs:
```diff
@@ -54,58 +52,61 @@ public string PackageFormatSelectorLabel
+            _settings.SaveToDisk();
         }
 
         private static bool IsSet(string value, bool defaultValue)
```
`IsSet(...)` is a poor name.  `Is-` methods should return bool, which is coincidentally true here but definitely untrue in the other `IsSet(...)` method below.  The method names should be something like `ReadValue(...)`.  Then adding the `defaultValue` named parameter at callsites would really improve readability.

dtivel commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceCredential.cs:
```diff
@@ -185,5 +186,46 @@ private static IEnumerable<string> ParseAuthTypeFilterString(string str)
+            combiner.AddObject(PasswordText);
+            combiner.AddObject(IsPasswordClearText);
+
+            return combiner.GetHashCode();
```
Is it worth memoizing the hash code, perhaps with `Lazy<int>`?  This type is immutable, so the hash code won't change.

PatoBeltran replied:
I was following the same behavior we have throughout the code where we don't memoize it, but I think it might be a good idea to memoize this since the type is immutable

dtivel commented on src/NuGet.Core/NuGet.Configuration/Utility/SettingsUtility.cs:
```diff
@@ -287,46 +339,63 @@ public static string GetDefaultPushSource(ISettings settings)
             }
 
-            return RevocationMode.Online;
+            return new List<string>();
```
In the general case I think `Enumerable.Empty<T>()` is better than returning `new List<T>()` (here and everywhere else in this file).

Call sites that expressly require `List<T>` (like [this](https://github.com/NuGet/NuGet.Client/pull/2370/files#diff-4dc98f9946385b4b19fd5d9223d01fc0R126)) should not regress performance since either way you're creating a new `List<T>`.  However, in any other place that does not require `List<T>`, you'd avoid wasteful instantiations.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Utility/ConfigurationContants.cs:
```diff
@@ -1,18 +1,22 @@
 {
     public static class ConfigurationConstants
     {
+        public static readonly string Add = "add";
```
It would be worth sorting constants by name.  Makes it easer to find and update.

dtivel commented on src/NuGet.Core/NuGet.Configuration/PackageSource/IPackageSourceProvider.cs:
```diff
@@ -8,17 +8,109 @@ namespace NuGet.Configuration
+        /// </summary>
+        /// <param name="name">Name of source to be searched for</param>
+        /// <returns>PackageSource that matches the given name. Null if none was found</returns>
+        PackageSource GetPackageSourceByName(string name);
```
Do any of these methods perform argument validation?  If so, add the exceptions they throw to XML doc comments.

dtivel commented on src/NuGet.Core/NuGet.Configuration/PackageSource/IPackageSourceProvider.cs:
```diff
@@ -8,17 +8,109 @@ namespace NuGet.Configuration
+        /// <summary>
+        /// Gets a list of all of the package sources
+        /// </summary>
+        /// <returns>List of all of the package sources</returns>
```
Enumerable

dtivel commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSource.cs:
```diff
@@ -133,6 +123,16 @@ public PackageSource(string source, string name, bool isEnabled)
             _hashCode = Name.ToUpperInvariant().GetHashCode() * 3137 + Source.ToUpperInvariant().GetHashCode();
         }
 
+        public SourceItem AsSourceItem()
```
Tests?

dtivel commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs:
```diff
@@ -61,62 +46,62 @@ private IEnumerable<PackageSource> LoadConfigurationDefaultSources(IEnumerable<P
+            var disabledSourcesSettings = disabledSourcesSection?.Items.Select(s => s as AddItem)
+                .Where(s => s != null);
 
+            var disabledSources = new HashSet<string>(disabledSourcesSettings?.GroupBy(setting => setting.Key).Select(group => group.First().Key) ?? new List<string>());
```
... ?? Enumerable.Empty<string>()

dtivel commented on src/NuGet.Core/NuGet.Configuration/Resources.resx:
```diff
@@ -120,9 +120,26 @@
+    <value>The item passed to the Update method cannot refer to a different item than the one being updated.</value>
+  </data>
+  <data name="CannotUpdateMachineWide" xml:space="preserve">
+    <value>Unable to update setting since it is in a machine wide NuGet.Config.</value>
```
machine-wide

dtivel commented on src/NuGet.Core/NuGet.Configuration/Resources.resx:
```diff
@@ -132,17 +149,35 @@
     <value>'{0}' must contain an absolute path '{1}'.</value>
   </data>
+  <data name="NoWritteableConfig" xml:space="preserve">
+    <value>There is no NuGet.Config that could be used for writting settings. Please create one at the desired location and restart the client.</value>
```
writing

dtivel commented on src/NuGet.Core/NuGet.Configuration/Resources.resx:
```diff
@@ -132,17 +149,35 @@
+  </data>
   <data name="Settings_FileName_Cannot_Be_A_Path" xml:space="preserve">
-    <value>Parameter 'fileName' to Settings must be just a fileName and not a path</value>
+    <value>Parameter 'fileName' to Settings must be just a fileName and not a path.</value>
```
The second `fileName` should be `file name`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs:
```diff
@@ -61,62 +46,62 @@ private IEnumerable<PackageSource> LoadConfigurationDefaultSources(IEnumerable<P
-            // Order the list so that they are ordered in priority order
-            var settingValues = sourceSettingValues.OrderByDescending(setting => setting.Priority);
+            var packageSourcesSection = Settings.GetSection(ConfigurationConstants.PackageSources);
+            var sources = packageSourcesSection?.Items.Select(s => s as SourceItem)
```
You can simplify this `.Select(...).Where(...)` casting pattern with `.OfType<SourceItem>()`.

This feedback applies throughout this file.

dtivel commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs:
```diff
@@ -239,38 +203,16 @@ private PackageSourceCredential ReadCredential(string sourceName)
-                            validAuthenticationTypesText: authenticationTypesText);
-                    }
-                }
+                return new PackageSourceCredential(sourceName, username, password, credentialsItem.IsPasswordClearText, authenticationTypes);
```
Why did you create locals for all `credentialsItem` properties except `IsPasswordClearText`?  Just curious.

PatoBeltran replied:
that was a ghost of a previous iteration haha. Nice catch!

dtivel commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs:
```diff
@@ -298,178 +240,472 @@ private PackageSourceCredential ReadCredentialFromEnvironment(string sourceName)
         }
 
-        public void SavePackageSources(IEnumerable<PackageSource> sources)
+        [Obsolete("DisablePackageSource(PackageSource source) is deprecated, please use DisablePackageSource(string name) instead.")]
```
"...deprecated. Please..." throughout the entire PR.

dtivel commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs:
```diff
@@ -298,178 +240,472 @@ private PackageSourceCredential ReadCredentialFromEnvironment(string sourceName)
                 {
-                    if (loadedPackageSources.Any(p => p.Equals(targetPackageSource)))
+                    var sourceMatching = loadedPackageSources.Any(p => p.Source.Equals(packageSource.Source, StringComparison.CurrentCultureIgnoreCase));
+                    var feedNameMatching = loadedPackageSources.Any(p => p.Name.Equals(packageSource.Name, StringComparison.CurrentCultureIgnoreCase));
```
`isSourceMatch` and `isFeedNameMatch`

dtivel commented on src/NuGet.Core/NuGet.Configuration/Resources.resx:
```diff
@@ -120,9 +120,26 @@
+  <data name="CannotUpdateMachineWide" xml:space="preserve">
+    <value>Unable to update setting since it is in a machine wide NuGet.Config.</value>
+  </data>
+  <data name="Error_CouldNotDisableSources" xml:space="preserve">
```
Where is this used?

PatoBeltran replied:
it doesn't seem to be used anymore, will delete

dtivel commented on src/NuGet.Core/NuGet.Configuration/Resources.resx:
```diff
@@ -132,17 +149,35 @@
   <data name="PackageSource_Invalid" xml:space="preserve">
     <value>The package source does not belong to the collection of available sources.</value>
   </data>
+  <data name="PropertyCannotBeNull" xml:space="preserve">
```
-`OrEmpty`

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/AbstractSettingSection.cs:
```diff
@@ -0,0 +1,157 @@
+                throw new ArgumentException(Resources.Error_MergeTwoDifferentSections);
+            }
+
+            foreach (var item in other.Items)
```
You can reduce block nesting by one level by changing the `foreach` expression to `other.Items.Where(item => item != null)`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingSection.cs:
```diff
@@ -0,0 +1,94 @@
+{
+    public abstract class SettingSection : SettingsGroup<SettingItem>, IEquatable<SettingSection>
+    {
+        public IReadOnlyCollection<SettingItem> Items => ChildrenSet.Select(c => c.Value).ToList();
```
Why do you need `.ToList()` on every get of `Items`?  Why not simply:

    public IReadOnlyCollection<SettingItem> Items => ChildrenSet.Values;

`Dictionary<SettingItem, SettingItem>.Values` implements `IReadOnlyCollection<SettingItem>`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingText.cs:
```diff
@@ -0,0 +1,67 @@
+
+        public override bool Equals(object other) => Equals(other as SettingText);
+
+        public override int GetHashCode() => Value.GetHashCode();
```
This will throw if `Value` is `null`, and `Value` is publicly settable to `null`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingsFile.cs:
```diff
@@ -0,0 +1,237 @@
+        /// <summary>
+        /// XML element for settings file
+        /// </summary>
+        private XDocument _xDocument { get; }
```
Why a private property?  Auto-implemented properties create a property and a private backing field.  Why not simply use a private read-only field?

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingsFile.cs:
```diff
@@ -0,0 +1,237 @@
+        /// Root element of configuration file.
+        /// By definition of a nuget.config, the root element has to be a 'configuration' element
+        /// </summary>
+        private NuGetConfiguration _rootElement { get; set; }
```
Use a private read-only field.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingsFile.cs:
```diff
@@ -0,0 +1,237 @@
+        /// </summary>
+        /// <param name="directoryPath">path to the directory where the file is</param>
+        public SettingsFile(string directoryPath)
+            : this(directoryPath, Settings.DefaultSettingsFileName, false)
```
Add a named parameter for `false`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingsFile.cs:
```diff
@@ -0,0 +1,237 @@
+        /// <param name="directoryPath">path to the directory where the file is</param>
+        /// <param name="fileName">name of config file</param>
+        public SettingsFile(string directoryPath, string fileName)
+            : this(directoryPath, fileName, false)
```
Add a named parameter for `false`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingsFile.cs:
```diff
@@ -0,0 +1,237 @@
+
+            if (string.IsNullOrEmpty(fileName))
+            {
+                throw new ArgumentNullException(nameof(fileName), Resources.Argument_Cannot_Be_Null_Or_Empty);
```
Both of these exceptions above should be `ArgumentException` not `ArgumentNullException`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingsFile.cs:
```diff
@@ -0,0 +1,237 @@
+
+        /// <remarks>
+        /// This method gives you a reference to the actual abstraction instead of a clone of it.
+        /// It should be used only when intended. For most purposed you should be able to use
```
purposes

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/AddItem.cs:
```diff
@@ -0,0 +1,160 @@
+            if (ReferenceEquals(this, other))
+            {
+                return true;
+            }
```
Why not replace 106-114 with:

    if (!Equals(other))
    {
        return false;
    }

Related to that, why is `Key` (in-)equality irrelevant to a `DeepEquals(...)` comparison?

PatoBeltran replied:
This is a great suggestion!... Key is not irrelevant in `DeepEquals`. `key` is an attribute, therefore the equality of keys is compared as part of the comparison between all attributes.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingBase.cs:
```diff
@@ -0,0 +1,77 @@
+        /// <summary>
+        /// Convenience method to add an Origin to an element and all its children when adding it in a group
+        /// </summary>
+        internal virtual void AddToOrigin(SettingsFile origin)
```
I don't see the value in having both a read/write `Origin` property and this method.  The XML doc comments say it's a convenience method, but looking through its call sites, I see no loss of convenience using a property setter over a method call (or visa versa).

PatoBeltran replied:
You are right, Origin should be read-only and only set through this method. The convenience of this method is that any given setting can override it and set it's children to origin as well.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingBase.cs:
```diff
@@ -0,0 +1,77 @@
+        public abstract bool DeepEquals(SettingBase other);
+
+        /// <summary>
+        /// Specifies if the setting is only an in-memory represantion or if
```
representation

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingBase.cs:
```diff
@@ -0,0 +1,77 @@
+        /// Compares two instances to see if they both match in their values for
+        /// the elements unique keys
+        /// </summary>
+        /// <remarks>Each setting defines it's unique keys,
```
its

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingBase.cs:
```diff
@@ -0,0 +1,77 @@
+        internal SettingBase(XNode node, SettingsFile origin)
+        {
+            Node = node ?? throw new ArgumentNullException(nameof(node));
+            Origin = origin ?? throw new ArgumentNullException(nameof(origin));
```
It seems absurd that the constructor disallows `null` yet the property setters allow `null`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/AbstractSettingSection.cs:
```diff
@@ -0,0 +1,157 @@
+            {
+                Debug.Assert(!currentSetting.IsAbstract());
+
+                if (currentSetting.Origin.IsMachineWide)
```
`Origin` can be `null`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs:
```diff
@@ -298,178 +240,472 @@ private PackageSourceCredential ReadCredentialFromEnvironment(string sourceName)
-                minPriority = existingSettings.Min(setting => setting.Priority);
+                try
+                {
+                    (Settings as Settings).AddOrUpdate(sourceSetting.Origin, ConfigurationConstants.DisabledPackageSources, new AddItem(name, "true"));
```
`sourceSetting.Origin` may be `null` and if it is `AddOrUpdate(...)` will throw a `NullReferenceException`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceProvider.cs:
```diff
@@ -298,178 +240,472 @@ private PackageSourceCredential ReadCredentialFromEnvironment(string sourceName)
+        private Dictionary<string, SourceItem> GetExistingSettingsLookup()
+        {
+            var sourcesSection = Settings.GetSection(ConfigurationConstants.PackageSources);
+            var existingSettings = sourcesSection?.Items.OfType<SourceItem>().Where(c => !c.Origin.IsMachineWide).ToList();
```
`Origin` may be `null`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/ClearItem.cs:
```diff
@@ -0,0 +1,35 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
```
Unused.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingItem.cs:
```diff
@@ -0,0 +1,95 @@
+                    {
+                        // Update or remove any existing item that has changed
+                        xElement.SetAttributeValue(attribute.Key, value);
+                        Origin.IsDirty = true;
```
`Origin` can be `null`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingItem.cs:
```diff
@@ -0,0 +1,95 @@
+
+        /// <remarks>
+        /// This method is internal because it updates directly the xElement behind this abstraction.
+        /// It should only be called whenever the underlaying config file is intended to be changed.
```
underlying

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingItem.cs:
```diff
@@ -0,0 +1,95 @@
+                if (xElement != null)
+                {
+                    xElement.SetAttributeValue(attribute.Key, attribute.Value);
+                    Origin.IsDirty = true;
```
`Origin` can be `null`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingSection.cs:
```diff
@@ -0,0 +1,94 @@
+            {
+                var currentChild = ChildrenSet[item];
+
+                if (currentChild.Origin.IsMachineWide)
```
`Origin` can be `null`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingFactory.cs:
```diff
@@ -0,0 +1,74 @@
+        {
+            var children = new List<T>();
+
+            var descendants = xElement.Elements().Select(d => Parse(d, origin)).OfType<T>().Distinct();
```
How does `Distinct()` dedupe here?  If I have a config file with multiple elements (but different attributes), what's the logic for deduplicating?

PatoBeltran replied:
It will dedup using the `Equals` function. This is intended because two items that match in `Equals` are referring to the same setting, so we only want to take the first and ignore the rest. This is the previous behavior of settings and this is what `Disctict()` does here.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingFactory.cs:
```diff
@@ -0,0 +1,74 @@
+
+                    return new AddItem(element, origin);
+                }
+                else if (string.Equals(element.Name.LocalName, ConfigurationConstants.Clear, StringComparison.OrdinalIgnoreCase))
```
Would it be more efficient normalizing case and then using a `switch` statement as opposed to a bunch of case-insensitive string comparisons?

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingFactory.cs:
```diff
@@ -0,0 +1,74 @@
+
+            foreach (var descendant in descendants)
+            {
+                if (descendant is ClearItem && canBeCleared)
```
A bool check seems more efficient than a type check, so I would tend to reverse the order of expressions.  However, I haven't tested the change.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/ParsedSettingSection.cs:
```diff
@@ -0,0 +1,36 @@
+        /// <remarks>
+        /// This constructor should only be used when the section is intended to be
+        /// added to a specific NuGetConfiguration. A ParsedSettingSection should not
+        /// be missing an Origin.
```
But there's nothing that enforces this.

The constructor above happily takes a `null` origin, this constructor doesn't set `Origin`, and the `Origin` property can be set to `null`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/NuGetConfiguration.cs:
```diff
@@ -0,0 +1,164 @@
+        internal NuGetConfiguration(SettingsFile origin)
+            : base()
+        {
+            var defaultSource = new SourceItem(NuGetConstants.FeedName, NuGetConstants.V3FeedUrl, protocolVersion: "3");
```
Shouldn't `3` be `PackageSourceProvider.MaxSupportedProtocolVersion.ToString()`?

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/NuGetConfiguration.cs:
```diff
@@ -0,0 +1,164 @@
+        internal NuGetConfiguration(XElement element, SettingsFile origin)
+            : base(element, origin)
+        {
+            if (element.Name != ElementName)
```
This is a case-sensitive comparision, which generally makes sense for XML, but elsewhere you perform case-insensitive comparsions (like on local names or [ElementName](https://github.com/NuGet/NuGet.Client/pull/2370/files#diff-5927462dd843bdc7c744cfd34c0b3ea7R156)).  What's the pattern or principle (if any)?

There should be tests for case (in-)sensitivity when matching element/attribute names.

PatoBeltran replied:
This should be a case-insensitive comparison like elsewhere. This was a copy from the old code. Good catch!

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/ConfigurationDefaults.cs:
```diff
@@ -62,15 +56,15 @@ public IEnumerable<PackageSource> DefaultPackageSources
                     _defaultPackageSources = new List<PackageSource>();
-                    var disabledPackageSources = _settingsManager.GetSettingValues(ConfigurationConstants.DisabledPackageSources);
-                    var packageSources = _settingsManager.GetSettingValues(ConfigurationConstants.PackageSources);
+                    var disabledPackageSources = _settingsManager.GetSection(ConfigurationConstants.DisabledPackageSources)?.Items.OfType<AddItem>() ?? new List<AddItem>();
```
`Enumerable.Empty<T>()` instead of `new List<T>()` across the PR.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/CredentialsItem.cs:
```diff
@@ -0,0 +1,298 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
```
Unused.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/UnknownItem.cs:
```diff
@@ -0,0 +1,235 @@
+                throw new ArgumentNullException(nameof(setting));
+            }
+
+            if (Origin.IsMachineWide)
```
`Origin` can be `null`.  Check all references for this problem.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SourceItem.cs:
```diff
@@ -0,0 +1,82 @@
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
```
Unused.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingsGroup.cs:
```diff
@@ -0,0 +1,139 @@
+    public abstract class SettingsGroup<T> : SettingElement, ISettingsGroup where T : SettingElement
+    {
+        // We use a dictionary instead of a hashset to be able to replace elements
+        protected Dictionary<T, T> ChildrenSet { get; private set; }
```
I don't understand this comment.  A hashset should work just fine.  You simply `Remove(...)` and `Add(...)` to replace.  It's not any more/less difficult than what you're currently doing with a `ContainsKey(...)`/`TryGetValue(...)` and `Remove(...)`.

A hashset would also be more consistent with the name you've chosen.

PatoBeltran replied:
`HashSet` does not have a `TryGetValue`, therefore the way to get the existing child would be iterating through the whole set. I'm basically trying to use a `Dictionary` as a set that also implements `TryGetValue`... Do you have a suggestion on how to make this better?

dtivel replied:
I see.  What you want is [this](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1.trygetvalue?view=netframework-4.7.2#System_Collections_Generic_HashSet_1_TryGetValue__0__0__).

I recommend updating the comment to say:

    Until HashSet<T>TryGetValue(...) is available, we will use a dictionary to enable efficient element retrieval.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/UnknownItem.cs:
```diff
@@ -0,0 +1,235 @@
+
+        protected override bool CanHaveChildren => true;
+
+        private Dictionary<SettingBase, SettingBase> MutableChildren { get; set; }
```
Why not `HashSet<SettingBase>`?

Also, don't use a private property.  Use a private field.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/UnknownItem.cs:
```diff
@@ -0,0 +1,235 @@
+
+        internal override XNode AsXNode()
+        {
+            if (Node != null && Node is XElement)
```
You don't need the explicit `null` check.  `is` includes a `null` check.  See https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/is.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/UnknownItem.cs:
```diff
@@ -0,0 +1,235 @@
+                return true;
+            }
+
+            return string.Equals(ElementName, other.ElementName, StringComparison.Ordinal) &&
```
Why isn't lines 161-171 simply:
```
if (!Equals(other))
{
    return false;
}
```
Also, why is this `ElementName` check ordinal, while the one on line 156 is ordinal case insensitive?

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/UnknownItem.cs:
```diff
@@ -0,0 +1,235 @@
+    {
+        public override string ElementName { get; protected set; }
+
+        public IReadOnlyList<SettingBase> Children => MutableChildren.Select(c => c.Value).ToList().AsReadOnly();
```
You don't have to call `.AsReadOnly()` as `List<T>` implements `IReadOnlyList<T>`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/UnknownItem.cs:
```diff
@@ -0,0 +1,235 @@
+
+        public IReadOnlyList<SettingBase> Children => MutableChildren.Select(c => c.Value).ToList().AsReadOnly();
+
+        public new IReadOnlyDictionary<string, string> Attributes => base.Attributes;
```
How about overriding a virtual base property instead of using `new`?

PatoBeltran replied:
The reason I used `new` is that `SettingElement` has an `Attributes` property that is internal. I want to expose that same property as `public`... Given that if I used `override` I could not change the `internal` to `public` I used a new. Why do you think `new` is not appropriate?

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/UnknownItem.cs:
```diff
@@ -0,0 +1,235 @@
+
+            var unknown = other as UnknownItem;
+
+            var otherChildren = new Dictionary<SettingBase, SettingBase>(unknown.MutableChildren);
```
`unknown` could be `null`.

PatoBeltran replied:
I will add a null check on `base.Update(...)` so all of the items are covered from this issue.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Utility/SettingsUtility.cs:
```diff
@@ -218,12 +269,13 @@ public static IReadOnlyList<string> GetFallbackPackageFolders(ISettings settings
-                .Select(setting => setting.Value)
+                .Select(f => f as AddItem)
+                .Where(f => f != null)
+                .Select(f => f.GetValueAsPath())
```
```
return fallbackValues
    .OfType<AddItem>()
    .Select(folder => folder.GetValueAsPath())
    .ToList();
```

dtivel commented on test/NuGet.Clients.Tests/NuGet.CommandLine.Test/NuGetConfigCommandTest.cs:
```diff
@@ -7,6 +7,8 @@
 using NuGet.Common;
 using Xunit;
+using System.Linq;
+using NuGet.Configuration;
```
Sort

dtivel commented on test/NuGet.Core.Tests/NuGet.Configuration.Test/TestUtils.cs:
```diff
@@ -18,5 +20,19 @@ public static void CreateConfigurationFile(string configurationPath, string mock
+
+        public static byte[] GetFileHash(string fileName)
+        {
+            var hashAlgorithm = SHA512.Create();
```
`HashAlgorithm` implements `IDisposable`.  You need to wrap this with a `using` statement.

dtivel commented on test/NuGet.Core.Tests/NuGet.Configuration.Test/FallbackFolderTests.cs:
```diff
@@ -185,7 +185,7 @@ public void GetFallbackPackageFolders_MultipleFoldersFromMultipleNuGetConfigs()
                     .Select(GetFileName);
 
-                var expected = new[] { "a", "b", "c", "d", "x", "y" };
+                var expected = new[] { "x", "y", "c", "d", "a", "b" };
```
Is this change correct?  Would returning the fallback folders in a different order affect package resolution?

PatoBeltran committed:
Address feedback

PatoBeltran commented:
@dtivel addressed most comments, missing adding additional tests. I also want to improve the error message experience for "Unable to parse config" to give a reason of what it found.

dtivel commented on src/NuGet.Core/NuGet.Configuration/PackageSource/PackageSourceCredential.cs:
```diff
@@ -80,6 +81,8 @@ public string Password
         /// <returns>True if credentials object is valid</returns>
         public bool IsValid() => !string.IsNullOrEmpty(Username) && !string.IsNullOrEmpty(PasswordText);
 
+        private Lazy<int> _hashCode { get; }
```
private readonly Lazy<int> _hashCode;

...and move to the top of the class (where fields typically go).

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/AddItem.cs:
```diff
@@ -0,0 +1,159 @@
+        public virtual string Value
+        {
+            get => Settings.ApplyEnvironmentTransform(Attributes[ConfigurationConstants.ValueAttribute]);
+            set => AddOrUpdateAttribute(ConfigurationConstants.ValueAttribute, value ?? string.Empty);
```
Why add `?? string.Empty`?

PatoBeltran replied:
Add items should always have `Key` and `Value` attributes, but `Value` can be empty... Therefore if a user set's it to null we don't want to delete it, but set it to empty.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/CredentialsItem.cs:
```diff
@@ -0,0 +1,331 @@
+
+        internal AddItem _password { get; set; }
+
+        internal AddItem _validAuthenticationTypes { get; set; }
```
This is a non-standard naming convention for properties.  The property names should be `Username`, `Password`, and `ValidAuthenticationTypes`, respectively.  However, I suspect that you made these private fields internal properties so that one instance of this class can access the non-public members of another instance of the class.  If so, it is sufficient to make these private fields.  Can they also be `readonly`?

PatoBeltran replied:
`_username` and `password` can be made `readonly` `_validAuthenticationTypes` can be deleted or re-created, therefore it cannot be `readonly`.

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/CredentialsItem.cs:
```diff
@@ -0,0 +1,331 @@
+                }
+                else if (string.Equals(item.Key, ConfigurationConstants.PasswordToken, StringComparison.OrdinalIgnoreCase))
+                {
+                    if (_password != null)
```
It looks like you're trying to block config files which have multiple password elements in the same credentials section.  This looks like new behavior.  Why enforce this on passwords but not other elements (like username and valid authentication types)?

Will this pattern be enforced more generally (across other settings)?

PatoBeltran replied:
We have never enforced this on code (i.e. failed reading) but we have never used more than one password if present, and it is implicit in our documentation and in the way we write the credentials that only one of this should be present. I could add validation also to make sure there is only one `_username` and one `_validAuthenticationTypes`.

What do you mean with this pattern being enforced more generally? `CredentialsItem` is making its own rules about how it should be constructed, each item can do the same. As for general add items in a section, [this](https://github.com/NuGet/NuGet.Client/pull/2370/files#diff-92bbf31cd5aa989dca530c722d6a93d1R2694) existing tests says that we should ignore duplicates instead of failing. Do you think is a better idea to follow this on `CredentialsItem` and just ignore any duplicates?

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingBase.cs:
```diff
@@ -0,0 +1,115 @@
+        /// <summary>
+        /// Convenience method to add an element to an origin
+        /// </summary>
+        /// <remarks>Each setting can override this method to inlcude any descendants to the origin</remarks>
```
include

dtivel commented on src/NuGet.Core/NuGet.Configuration/Settings/SettingsFile.cs:
```diff
@@ -177,6 +184,8 @@ internal bool TryGetSection(string sectionName, out SettingSection section)
 
         internal static void ConnectSettingsFilesLinkedList(IList<SettingsFile> settingFiles)
         {
+            settingFiles[0].Priority = settingFiles.Count;
```
What if `settingFiles` is empty?

PatoBeltran committed:
Add missing tests and address feedback
